File: interview70/twoSum.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
    public:
        vector<int> twoSum(vector<int>& nums, int target) {
            unordered_map<int, int> m {};
            vector<int> result;
            int remainder;
            

            for (int i = 0; i < (int) nums.size(); ++i) {
                remainder = target - nums[i];
                if (m.find(remainder) == m.end()) {
                    m.insert({ nums[i], i });
                } 
                else {
                    result.push_back(m[remainder]);
                    result.push_back(i);
                }
            }

            return result;
        }

        void showVector(vector<int> &nums) {
            for (int each : nums) {
                cout << each << " ";
            }
        }
};


int main(int argc, char* argv[]) {
    vector<int> v = {2,7,11,15};
    int target = 9;
    Solution s;
    vector<int> result = s.twoSum(v, target);
    s.showVector(result);

    return 0;
}

File: interview70/TwoSum.java
package interview70;

import java.util.HashMap;
import java.util.Map;

public class TwoSum {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer,Integer> map = new HashMap<>(); 

        for (int i = 0; i < nums.length; ++i) {
            int complement = target - nums[i];
            if (map.containsKey(complement)) {
                return new int[] { map.get(complement), i };
            }
            map.put(nums[i], i);
        }

        throw new IllegalArgumentException("No solution");
    }

    static public void showArray(int[] arr) {
        for (int each : arr) {
            System.out.printf("%d ", each);
        }
    } 

    public static void main(String[] args) {
        TwoSum twosum = new TwoSum();
        int[] arr = {2,11,7,15};
        int target = 9;
        int[] result = twosum.twoSum(arr, target);
        TwoSum.showArray(result);

    }
}
